gpu-compute,tests: Move GPU tests to testlib

BobbyRBruce opened the pull request:
A new host tag `gcn_gpu` has been added. This allows for selection of those GPU tests which depend upon the gcn-gpu docker image to run.

In addition to this, the square GPU tests has been moved to the CI tests. This ensures some GPU code is compiled and run on every PR.

Change-Id: Ic1a3e3a5c14566c5a53aeb74ab6d0725a1310fff

BobbyRBruce committed:
gpu-compute,tests: Move GPU tests to testlib

A new host tag `gcn_gpu` has been added. This allows for selection of
those GPU tests which depend upon the gcn-gpu docker image to run.

In addition to this, the square GPU tests has been moved to the CI
tests. This ensures some GPU code is compiled and run on every PR.

Change-Id: Ic1a3e3a5c14566c5a53aeb74ab6d0725a1310fff

BobbyRBruce commented:
I'm not sure if this is really any better, but just what I came up with trying to get the GPU tests into test lib. Due to the docker container requirement they are still a special case.

mattsinc commented on .github/workflows/ci-tests.yaml:
```diff
@@ -221,6 +221,32 @@ jobs:
+            - name: Build VEGA_X86/gem5.opt
+              run: scons build/VEGA_X86/gem5.opt -j`nproc`
+
+            # Run the GPU tests.
```
How do I understand what tests this runs?  It does not seem to correspond to any of the other file names in this commit ...

BobbyRBruce replied:
The code that follows comment this calls our TestLib framework:

```shell
./main.py run  --skip-build -vvv -t $(nproc) --host gcn_gpu gem5/gpu
```

The important part here is the "gem5/gpu" at the end. That means it's running the tests defined in the "tests/gem5/gpu" directory. For example: https://github.com/gem5/gem5/blob/fe8b2b24eb9b08c07badd6f08fdeedc32db5a5c1/tests/gem5/gpu/test_gpu_apu_se.py. This script show what script gem5 is running and with what parameters using the Vega binary we built in the previous step.

The only snag with this implementation is needing to pass `--host gcn_gpu` to testlib to run the tests. This is because we need to compile VEGA and run the tests in the special docker container and I use this flag to ensure this happens.

mattsinc commented on .github/workflows/daily-tests.yaml:
```diff
@@ -80,6 +80,30 @@ jobs:
+            - name: Build VEGA_X86/gem5.opt
+              run: scons build/VEGA_X86/gem5.opt -j`nproc`
+
+            - name: Run Testlib GPU Tests
```
Same question here

BobbyRBruce replied:
My answer here is similar to the above but the only difference here is passing `--length=long` to state we want to run the Daily/Long tests found in the "tests/gem5/gpu" directory.

BobbyRBruce committed:
tests,ext,gpu-compute: Fix `gcn_gpu` host tag

This was not implemented correctly resulting in the GPU tests being run
even when the host tag is not set.

Change-Id: I166fb975dade08114f288f638fc7f534dd6f4711

mattsinc reviewed (approved):
Thanks @BobbyRBruce , I don't feel like I have a strong grasp on this code, but based on your reply seems ok ..
